### PR TITLE
EthereumProvider and the related refactoring

### PR DESCRIPTION
**Pull request by nivida**
- [x] Port Web3 to es6.
- [x] Refactor the architecture to more SOLID & clean code.
- [x] Improve provider handling (Web3 should use internally always the same interface).
- [x] Each RPC method should be implemented once. 
- [x] Method calls should be delegated with the JS Proxy object instead of adding them on construction.
- [x] Create AbstractWeb3Module class and remove ```core.packageInit()```.
- [x] Simplify the internal subscription handling.
- [x] Improve the documentation.
- [x] Fix the log subscription.
- [x] Write unit tests with Jest and move them to the related module folders.
- [x] defaultBlock, defaultAccount, timeoutBlock and confirmationBlock should be configurable for each module over the constructor.
- [x] Refactor ENS module (dependency inversion and not injection, EIP 1577)
- [x] Set up rollup for each module
- [x] Set up eslint & prettier
- [x] Update travis CI 
- [x] Improve web3-utils & web3-core-helpers exports

### CODE REVIEW

File: packages/web3-core-method/src/commands/SignMessageCommand.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * @constructor
+ */
+function SignMessageCommand(messageSigner) {
+    this.messageSigner = messageSigner;
+}
+
+/**
+ * Executes the SignMessageCommand and returns the signed message
+ *
```
Comment (nivida) [inline packages/web3-core-method/src/commands/SignMessageCommand.js]: add ```@method execute```

File: packages/web3-core-method/lib/commands/AbstractSendMethodCommand.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+ * @returns {Boolean}
+ */
+AbstractSendMethodCommand.prototype.isGasPriceDefined = function (parameters) {
+    return _.isObject(parameters[0]) && typeof parameters[0].gasPrice !== 'undefined';
+};
+
+/**
+ * Returns the current gasPrice of the connected node
+ *
+ * @param {AbstractProviderAdapter | EthereumProvider} provider
```
Comment (nivida) [inline packages/web3-core-method/lib/commands/AbstractSendMethodCommand.js]: remove spaces

File: packages/web3-core-method/lib/models/AbstractMethodModel.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        parameters: parameters
+    }
+};
+
+/**
+ * Determines if the JSON-RPC method is sign.
+ *
+ * @method isSign
+ *
+ * @returns {boolean}
```
Comment (nivida) [inline packages/web3-core-method/lib/models/AbstractMethodModel.js]: uppercase

File: packages/web3-core-method/lib/models/AbstractMethodModel.js [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+AbstractMethodModel.prototype.isSign = function () {
+    return this.rpcMethod === 'eth_sign';
+};
+
+/**
+ * Determines if the JSON-RPC method is sendTransaction
+ *
+ * @method isSendTransaction
+ *
+ * @returns {boolean}
```
Comment (nivida) [inline packages/web3-core-method/lib/models/AbstractMethodModel.js]: uppercase

File: gulpfile.js [hunk 5]
```diff
@@ -71,27 +67,13 @@ var packages = [{
...
-    expose: 'Web3WsProvider',
-    src: './packages/web3-providers-ws/src/index.js',
-    ignore: ['websocket']
+    fileName: 'web3-core-providers',
+    expose: 'Web3Providers',
+    src: './packages/web3-core-providers/src/index.js'
 }, {
     fileName: 'web3-core-subscriptions',
-    expose: 'Web3Subscriptions',
+    expose: 'Web3Subscription',
```
Comment (frozeman) [inline gulpfile.js]: Why you renamed it?

File: package.json [hunk 6]
```diff
@@ -1,94 +1,101 @@
...
+    "authors": [
+        {
+            "name": "Samuel Furter",
+            "email": "samuel@ethereum.org",
+            "homepage": "https://github.com/nivida"
+        },
+        {
+            "name": "Fabian Vogelsteller",
+            "email": "fabian@ethereum.org",
+            "homepage": "http://frozeman.de"
```
Comment (frozeman) [inline package.json]: can you change mine to `https://github.com/frozeman` :)

File: packages/web3-bzz/src/index.js [hunk 7]
```diff
@@ -16,73 +16,28 @@
...
+    /**
+     * Returns the Bzz object
+     *
+     * @method createBzz
+     *
+     * @param {Object|String} provider
+     *
+     * @returns {Bzz}
+     */
+    createBzz: function (provider) {
```
Comment (frozeman) [inline packages/web3-bzz/src/index.js]: why you added this method?
Does this now expose Bzz.createBzz()?

Packages should always be initiated as follows: `new Bzz(provider)`

File: packages/web3-core-helpers/src/formatters.js [hunk 8]
```diff
@@ -43,13 +43,24 @@ var isPredefinedBlockNumber = function (blockNumber) {
...
+/**
+ * Determines if it should use the default block by the given package or not,
+ * will map 'genesis' and 'earlist' to '0x0' and runs the inputBlockNumberFormatter.
+ *
+ * @param {String|Number} blockNumber
+ * @param {AbstractWeb3Object} web3Package
+ *
+ * @returns {String}
+ */
+var inputDefaultBlockNumberFormatter = function (blockNumber, web3Package) {
```
Comment (frozeman) [inline packages/web3-core-helpers/src/formatters.js]: Whats the web3Package here? the instance of the current one? Whats you thought being this, vs. using the `this` reference?

Comment (frozeman) [reply packages/web3-core-helpers/src/formatters.js]: Also the question would be here that the `this` can be the `Eth` module, rather than the whole `web3` instance. This formatter could be attached to all sub packages, and they need to work independent from the `web3` package.

File: packages/web3-core-helpers/src/formatters.js [hunk 9]
```diff
@@ -106,43 +117,50 @@ var _txInputFormatter = function (options){
...
+        if (!options.from) {
+            options.from = web3Package.defaultAccount;
+        }
 
         if (!options.from && !_.isNumber(options.from)) {
             throw new Error('The send transactions "from" field must be defined!');
         }
 
-        options.from = inputAddressFormatter(options.from);
+        options.from = (options.from);
```
Comment (frozeman) [inline packages/web3-core-helpers/src/formatters.js]: Why did you remove the `inputAddressFormatter` here?

File: packages/web3-core-helpers/src/formatters.js (hunk 8 above)
Comment (frozeman) [reply packages/web3-core-helpers/src/formatters.js]: The idea here is that e.g. the `defaultBlock` can be changed globally on your package instance (`web3`, `eth`, etc) and these functions should always use whatever the developer set to be the default on his package instance.

File: packages/web3-core-method/README.md [hunk 10]
```diff
@@ -29,17 +29,77 @@ This will expose the `Web3Method` object on the window object.
...
+
+
+// Create the MyMethoModelFactory object
+/**
+ * @param {Object} utils
+ * @param {Object} formatters
+ * 
+ * @constructor
+ */
+function MyMethodModelFactory(utils, formatters) {
```
Comment (frozeman) [inline packages/web3-core-method/README.md]: not sure if the prefix `My` is necessary here.

File: packages/web3-bzz/README.md [hunk 11]
```diff
@@ -22,16 +21,16 @@ npm run-script build-all
...
 
 ## Usage
 
 ```js
 // in node.js
-var Web3Bzz = require('web3-bzz');
+var bzzPackage = require('web3-bzz');
 
-var bzz = new Web3Bzz('http://swarm-gateways.net');
+var bzz = bzzPackage.createBzz('http://swarm-gateways.net');
```
Comment (frozeman) [inline packages/web3-bzz/README.md]: Please instantiate using `new Bzz('....')`

File: packages/web3-core-helpers/src/formatters.js [hunk 12]
```diff
@@ -43,13 +43,24 @@ var isPredefinedBlockNumber = function (blockNumber) {
...
+ * Determines if it should use the default block by the given package or not,
+ * will map 'genesis' and 'earlist' to '0x0' and runs the inputBlockNumberFormatter.
+ *
+ * @param {String|Number} blockNumber
+ * @param {AbstractWeb3Object} web3Package
+ *
+ * @returns {String}
+ */
+var inputDefaultBlockNumberFormatter = function (blockNumber, web3Package) {
+    if (blockNumber === undefined || blockNumber === null) {
```
Comment (frozeman) [inline packages/web3-core-helpers/src/formatters.js]: add check if `web3Package` exists

Comment (frozeman) [reply packages/web3-core-helpers/src/formatters.js]: `web3Package ` -> `parentModuleInstance`

File: packages/web3-core-method/src/models/TransactionConfirmationModel.js [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        set: function () {},
+        enumerable: true
+    });
+
+    /**
+     * Defines accessors for TIMEOUTBLOCK
+     *
+     * Created empty setter that it acts like a constant.
+     */
+    Object.defineProperty(this, 'TIMEOUTBLOCK', {
```
Comment (frozeman) [inline packages/web3-core-method/src/models/TransactionConfirmationModel.js]: `eth.options.transactionBlockTimeout`

File: packages/web3-core-method/src/models/TransactionConfirmationModel.js [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        set: function () {},
+        enumerable: true
+    });
+
+    /**
+     * Defines accessors for CONFIRMATIONBLOCKS
+     *
+     * Created empty setter that it acts like a constant.
+     */
+    Object.defineProperty(this, 'CONFIRMATIONBLOCKS', {
```
Comment (frozeman) [inline packages/web3-core-method/src/models/TransactionConfirmationModel.js]: `eth.options.transactionConfirmationBlocks`

File: packages/web3-core-method/src/models/TransactionConfirmationModel.js [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+function TransactionConfirmationModel() {
+    this.confirmations = [];
+    this.timeoutCounter = 0;
+
+    /**
+     * Defines accessors for POLLINGTIMEOUT. This is the average block time (seconds) * TIMEOUTBLOCK
+     *
+     * Created empty setter that it acts like a constant.
+     */
+    Object.defineProperty(this, 'POLLINGTIMEOUT', {
```
Comment (frozeman) [inline packages/web3-core-method/src/models/TransactionConfirmationModel.js]: ``eth.options.transactionPollingTimeout``

File: packages/web3-core-method/src/models/methods/block/GetUncleMethodModel.js [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+GetUncleMethodModel.prototype = Object.create(AbstractMethodModel.prototype);
+GetUncleMethodModel.prototype.constructor = GetUncleMethodModel;
+
+/**
+ * This method will be executed before the RPC request.
+ *
+ * @method beforeExecution
+ *
+ * @param {AbstractWeb3Object} web3Package
```
Comment (frozeman) [inline packages/web3-core-method/src/models/methods/block/GetUncleMethodModel.js]: `AbstractWeb3Object ` -> `{AbstractWeb3Module} moduleInstance`

**Comment by filips123** [comment]
@nivida 

Package `ethers` (which causes issues in #2172) was recently updated to `4.0.21` so it is again a bit outdated. Why not require it with `^4.0.0` or something like this so it will be always updated?

Also, can this change be added to separate PR and published faster. It is currently impossible to build Web3 with Webpack without manually requiring latest `ethers` version.

**Comment by nivida** [comment]
@filips123 I define fixed versions for external dependencies because of security reasons but will update ethers to the latest version in this branch before I merge it. I will go back to a monthly patch release cycle after the v1.0.0-beta.38 release and then the ethers version and other dependencies will get updated each month.

**Comment by filips123** [comment]
When will this changes be released? Would it be faster to add and merge `ethers` update in different PR?

**Comment by nivida** [comment]
I‘m currently fixing and improving the typings and did create a 1.0.0-beta.38-rc0 tag and asked severall projects to test it and also completly tested it by myself. This means that it should get released within the next week. 

Feel free to test the 1.0.0-beta.38-rc0 tag and to give some feedback from the tests.

**Comment by levino** [comment]
Before you merge this: Please open a PR to DT to [remove](https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package) the type definitions from there.

**Comment by Keith-CY** [comment]
@nivida sorry for that, new commit proposed https://github.com/ethereum/web3.js/pull/2207

**Comment by filips123** [comment]
@nivida This PR is now merged. When will new release be published?

**Comment by levino** [comment]
OMG. I has been merged!! :)

**Comment by EddieOne** [comment]
How do we use these features?

**Comment by wbt** [comment]
As a request from those who [may need to go through history of changes later](https://github.com/trufflesuite/truffle/issues/1912#issuecomment-508133730) as part of debugging relevant issues, can we please for the future try to move toward [smaller](https://opensource.zalando.com/blog/2017/10/small-pull-requests/) [pull](https://sourcediving.com/a-practical-guide-to-small-and-easy-to-review-pull-requests-a7f04a01d5d5) [requests](https://medium.com/bestseller-e-commerce-technology/5-small-things-to-improve-your-pull-request-3210d0722c09) than monsters like this one?
